### PR TITLE
feat: Adds sortable table detail page

### PR DESCRIPTION
**Pull request by Golodhros**
### Summary of Changes
Adds ability to sort the table columns on the Table detail page

core_fact_rides_-_Amundsen_Table_Details

### Tests
Added tests

### CheckList

Make sure you have checked **all** steps below to ensure a timely review.

- [ ] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [ ] PR includes a summary of changes, including screenshots of any UI changes.
- [ ] PR adds unit tests, updates existing unit tests, **OR** documents why no test additions or modifications are needed.
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public python functions and the classes in the PR contain docstrings that explain what it does
- [ ] PR passes all tests documented in the [developer guide](https://github.com/lyft/amundsenfrontendlibrary/blob/master/docs/developer_guide.md#testing)

### CODE REVIEW

File: amundsen_application/static/js/components/ColumnList/index.tsx [hunk 1]
```diff
@@ -29,6 +29,9 @@ import {
 
 import './styles.scss';
 
+export type SortDirection = 'asc' | 'desc';
```
Comment (ttannis) [inline amundsen_application/static/js/components/ColumnList/index.tsx:32]: nit: I have a preference for the full words over abbreviations so that nobody misinterprets what things are meant to do. For example, here I have to use the context clue of `SortDirection` to understand that `desc` means `descending` as opposed to maybe `description`. So in favor of `ascending` and `descending`

File: amundsen_application/static/js/pages/TableDetailPage/index.tsx [hunk 2]
```diff
@@ -105,13 +126,26 @@ const ErrorMessage = () => {
...
-  TableDetailProps & RouteComponentProps<any>
+  TableDetailProps & RouteComponentProps<any>,
+  StateProps
 > {
   private key: string;
 
   private didComponentMount: boolean = false;
 
+  // @ts-ignore
+  state = {
```
Comment (ttannis) [inline amundsen_application/static/js/pages/TableDetailPage/index.tsx:145]: What issue does this cause that we have to ignore it? Does the fact that we're not initializing in a `constructor` method have any effect?

File: amundsen_application/static/js/pages/TableDetailPage/index.tsx [hunk 3]
```diff
@@ -65,8 +66,28 @@ import './styles.scss';
...
+  },
+};
+const SORT_OPTIONS = [
+  {
+    name: 'Table Default',
+    type: 'sort_order',
+  },
+  {
+    name: 'Usage Count',
+    type: 'usage',
```
Comment (ttannis) [inline amundsen_application/static/js/pages/TableDetailPage/index.tsx:86]: Does this work right now as is? This looks like it works by needing a `usage` attribute on the column metadata, but usage is on the column stats.

Given this implementation -- unless I've missed some recent changes -- we will have to add logic in https://github.com/amundsen-io/amundsenfrontendlibrary/blob/master/amundsen_application/api/utils/metadata_utils.py#L68 to pull the column usage object from the stats list (so that it doesnt show up on the stats UI), and add its value directly on the column object under the name `usage`.

File: amundsen_application/static/js/components/ColumnList/index.tsx (hunk 1 above)
Comment (Golodhros) [reply amundsen_application/static/js/components/ColumnList/index.tsx:32]: Oh, actually I was working on exactly that!
Using an enum, check the updated PR!

File: amundsen_application/static/js/pages/TableDetailPage/index.tsx (hunk 2 above)
Comment (Golodhros) [reply amundsen_application/static/js/pages/TableDetailPage/index.tsx:145]: Not sure!
I don't think so, usually this method is preferred.

File: amundsen_application/static/js/pages/TableDetailPage/index.tsx (hunk 3 above)
Comment (Golodhros) [reply amundsen_application/static/js/pages/TableDetailPage/index.tsx:86]: Oh, I do some rework of the data inside Column list o add that info into the data itself.

File: amundsen_application/static/js/pages/TableDetailPage/index.tsx (hunk 2 above)
Comment (Golodhros) [reply amundsen_application/static/js/pages/TableDetailPage/index.tsx:129]: Actually, it doesn't fail anymore. Updating it!

File: amundsen_application/static/js/components/common/Table/index.tsx [hunk 4]
```diff
@@ -8,8 +8,14 @@ import { UpIcon, DownIcon } from '../SVGIcons';
 
 import './styles.scss';
 
-type TextAlignmentValues = 'left' | 'right' | 'center';
+// export type SortDirection = 'asc' | 'desc';
```
Comment (allisonsuarez) [inline amundsen_application/static/js/components/common/Table/index.tsx:11]: Why are these commented out?

Comment (Golodhros) [reply amundsen_application/static/js/components/common/Table/index.tsx:11]: I will implement it shortly!

File: amundsen_application/static/js/components/ColumnList/index.spec.tsx [hunk 5]
```diff
@@ -96,6 +113,27 @@ describe('ColumnList', () => {
...
+        it('should sort the data by that value', () => {
+          const { wrapper } = setup({
+            columns,
+            sortBy: {
+              key: 'usage',
+              direction: SortDirection.descending,
+              name: 'Usage',
+            },
+          });
+          const expected = 'simple_column_name_timestamp';
```
Comment (allisonsuarez) [inline amundsen_application/static/js/components/ColumnList/index.spec.tsx:127]: So here are you testing if it was sorted correctly by checking just the first item to see if its the one that is supposed to be first?

File: amundsen_application/static/js/components/ColumnList/index.tsx [hunk 6]
```diff
@@ -153,17 +197,26 @@ const ColumnList: React.FC<ColumnListProps> = ({
         name: item.name,
         database,
       },
-      usage: hasItemStats ? item.stats[0].stat_val : '',
+      sort_order: item.sort_order,
+      usage: hasItemStats ? +item.stats[0].stat_val : null,
```
Comment (ttannis) [inline amundsen_application/static/js/components/ColumnList/index.tsx:201]: I don't think the api guarantees the column usage will be the first statistic, so perhaps we should have a helper method to intentionally search for the `column_usage` statistic and extract the value.

File: amundsen_application/static/js/components/ColumnList/index.tsx [hunk 7]
```diff
@@ -153,17 +197,26 @@ const ColumnList: React.FC<ColumnListProps> = ({
...
+      name: item.name,
       isEditable: item.is_editable,
       editText,
       editUrl,
       index,
     };
   });
   const statsCount = formattedData.filter((item) => !!item.stats).length;
-  const hasStats = statsCount >= SHOW_STATS_THRESHOLD;
+  const hasStats =
```
Comment (ttannis) [inline amundsen_application/static/js/components/ColumnList/index.tsx:212]: nit: Is this more accurately `hasUsage`?

File: amundsen_application/static/js/pages/TableDetailPage/index.tsx (hunk 3 above)
Comment (ttannis) [reply amundsen_application/static/js/pages/TableDetailPage/index.tsx]: Just noticed it. Left a comment about it below.

**Comment by ttannis** [comment]
Looks like tests are failing because of amundsencommon? cc @allisonsuarez since I think you were working on it. On my end only concern is my comment about reliable logic to extract the column usage from the stats.

**Comment by feng-tao** [comment]
> Looks like tests are failing because of amundsencommon? cc @allisonsuarez since I think you were working on it. On my end only concern is my comment about reliable logic to extract the column usage from the stats.

let me take a look at the code as well. I was raising a question: could we sort based on a metric which defined by usage? We name column usage but other OSS could name it differently.

Also is @ttannis 's concern extracting column usage in stats requiring certain hack? do you think if we put query / column usage as separate node helps?

File: amundsen_application/static/js/components/ColumnList/index.tsx (hunk 6 above)
Comment (Golodhros) [reply amundsen_application/static/js/components/ColumnList/index.tsx:201]: Good call!
Right now this will work, but as we add more stats it will definitely break.
I will add something now.

File: amundsen_application/static/js/components/ColumnList/index.tsx (hunk 7 above)
Comment (Golodhros) [reply amundsen_application/static/js/components/ColumnList/index.tsx:212]: Yep, will change

File: amundsen_application/static/js/components/ColumnList/index.tsx (hunk 6 above)
Comment (feng-tao) [reply amundsen_application/static/js/components/ColumnList/index.tsx:201]: +1, currently we have single stat at Lyft, but we will ingest more stat later. I think we should either define the query / metric name in config for sorting or redefine the usage not part of stat. I wonder whether we could choose the former?

File: amundsen_application/static/js/components/ColumnList/index.tsx [hunk 8]
```diff
@@ -76,6 +90,50 @@ type ExpandedRowProps = {
 };
 
 const SHOW_STATS_THRESHOLD = 1;
+const USAGE_STAT_TYPE = 'column_usage';
```
Comment (Golodhros) [inline amundsen_application/static/js/components/ColumnList/index.tsx:93]: This would probably be the item to move to the config, however I am wary of creating that abstraction now, without knowing more about what it is to come.

Comment (ttannis) [reply amundsen_application/static/js/components/ColumnList/index.tsx:93]: I can understand that. I am alright with it as it since the community will be able to follow existing examples to make this more flexible if/when needed.

**Comment by ttannis** [comment]
> > Looks like tests are failing because of amundsencommon? cc @allisonsuarez since I think you were working on it. On my end only concern is my comment about reliable logic to extract the column usage from the stats.
> 
> let me take a look at the code as well. I was raising a question: could we sort based on a metric which defined by usage? We name column usage but other OSS could name it differently.
> 
> Also is @ttannis 's concern extracting column usage in stats requiring certain hack? do you think if we put query / column usage as separate node helps?

I wouldn't call it a hack, rather extra logic to explicitly search the stats array for `column_usage`. Having it as a separate node would make it such that we don't need that extra logic, but I still believe your original perspective that column usage is a statistic in the metadata is fair. The case here is that the UI design has chosen to display usage differently from other column statistics, so I think it is appropriate for the frontend to have to have go parse the data because of how our designs have chosen to display them. 

In short, no strong opinion on changing it!

**Review by ttannis — APPROVED** [review]
Pending passing CI checks.

File: amundsen_application/static/js/components/ColumnList/index.tsx [hunk 9]
```diff
@@ -153,17 +212,26 @@ const ColumnList: React.FC<ColumnListProps> = ({
...
       isEditable: item.is_editable,
       editText,
       editUrl,
       index,
     };
   });
   const statsCount = formattedData.filter((item) => !!item.stats).length;
-  const hasStats = statsCount >= SHOW_STATS_THRESHOLD;
+  const hasUsageStat =
+    getTableSortCriterias().usage && statsCount >= SHOW_STATS_THRESHOLD;
```
[the remark this answers is not in the document]
Comment (Golodhros) [reply amundsen_application/static/js/components/ColumnList/index.tsx:228]: If no element on the data has usage, we won't show the whole column, right.

File: amundsen_application/static/js/config/config-default.ts [hunk 10]
```diff
@@ -162,6 +162,18 @@ const configDefault: AppConfig = {
...
+        sort_order: {
+          name: 'Table Default',
+          key: 'sort_order',
+          direction: SortDirection.ascending,
+        },
+        name: {
+          name: 'Alphabetical',
+          key: 'name',
+          direction: SortDirection.descending,
+        },
```
[the remark this answers is not in the document]
Comment (Golodhros) [reply amundsen_application/static/js/config/config-default.ts:175]: Yep, that change is on the private repo, only for us right now.

**Comment by Golodhros** [comment]
> > > Looks like tests are failing because of amundsencommon? cc @allisonsuarez since I think you were working on it. On my end only concern is my comment about reliable logic to extract the column usage from the stats.
> > 
> > 
> > let me take a look at the code as well. I was raising a question: could we sort based on a metric which defined by usage? We name column usage but other OSS could name it differently.
> > Also is @ttannis 's concern extracting column usage in stats requiring certain hack? do you think if we put query / column usage as separate node helps?
> 
> I wouldn't call it a hack, rather extra logic to explicitly search the stats array for `column_usage`. Having it as a separate node would make it such that we don't need that extra logic, but I still believe your original perspective that column usage is a statistic in the metadata is fair. The case here is that the UI design has chosen to display usage differently from other column statistics, so I think it is appropriate for the frontend to have to have go parse the data because of how our designs have chosen to display them.
> 
> In short, no strong opinion on changing it!

Done!
